target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+		.rset = _rset,			\
+	}
+
+static const struct rtcl_rtab_set rtcl_rtab_set[SOC_COUNT][CLK_COUNT] = {
+	{
+		RTAB_SET(rtcl_838x_cpu_reg_set),
+		RTAB_SET(rtcl_838x_mem_reg_set),
+		RTAB_SET(rtcl_838x_lxb_reg_set)
+	}, {
+		RTAB_SET(rtcl_839x_cpu_reg_set),
+		RTAB_SET(rtcl_839x_mem_reg_set),
+		RTAB_SET(rtcl_839x_lxb_reg_set)
+	}
+};
+
+#define ROUND_SET(_min, _max, _step)	\
+	{				\
+		.min = _min,		\
+		.max = _max,		\
+		.step = _step,		\
+	}
+
+struct rtcl_round_set {
+	unsigned long min;
+	unsigned long max;
+	unsigned long step;
+};
+
+static const struct rtcl_round_set rtcl_round_set[SOC_COUNT][CLK_COUNT] = {
+	{
+		ROUND_SET(300000000, 625000000, 25000000),
+		ROUND_SET(200000000, 375000000, 25000000),
+		ROUND_SET(100000000, 200000000, 25000000)
+	}, {
+		ROUND_SET(400000000, 850000000, 25000000),
+		ROUND_SET(100000000, 400000000, 25000000),
+		ROUND_SET(50000000, 200000000, 50000000)
+	}
+};
+
+static const int rtcl_divn3[] = { 2, 3, 4, 6 };
+static const int rtcl_xdiv[] = { 2, 4, 2 };
+
+/*
+ * module data structures
+ */
+
+#define COMPAT_RTL838X "realtek,rtl8380-clock"
+#define COMPAT_RTL839X "realtek,rtl8390-clock"
+
+#define CLK_INFO(_idx, _name, _pname, _dname, _dtrange)	\
+	{						\
+		.idx = _idx,				\
+		.name = _name,				\
+		.parent_name = _pname,			\
+		.display_name = _dname,			\
+		.dtrange = _dtrange			\
+	}
+
+struct rtcl_clk_info {
+	unsigned int idx;
+	const char *name;
+	const char *parent_name;
+	const char *display_name;
+	const char *dtrange;
+};
+
+struct rtcl_clk {
+	unsigned int idx;
+	unsigned long min;
+	unsigned long max;
+	unsigned long startup;
+	struct clk_hw hw;
+};
+
+static const struct rtcl_clk_info rtcl_clk_info[CLK_COUNT] = {
+	CLK_INFO(CLK_CPU, "cpu_clk", "ref_clk", "CPU", "cpu-range"),
+	CLK_INFO(CLK_MEM, "mem_clk", "ref_clk", "MEM", "mem-range"),
+	CLK_INFO(CLK_LXB, "lxb_clk", "ref_clk", "LXB", "lxb-range")
+};
+
+struct rtcl_dram {
+	int type;
+	int buswidth;
+};
+
+struct rtcl_sram {
+	unsigned long vbase;
+	int *pmark;
+};
+
+struct rtcl_ccu {
+	int soc;
+	struct rtcl_sram sram;
+	struct rtcl_dram dram;
+	bool dynamic;
+	spinlock_t lock;
+	struct platform_device *pdev;
+	struct device_node *np;
+	struct rtcl_clk clks[CLK_COUNT];
+};
+
+struct rtcl_ccu *rtcl_ccu;
+
+#define rtcl_hw_to_clk(_hw) container_of(_hw, struct rtcl_clk, hw)
+
+/*
+ * SRAM relocatable assembler functions
+ */
+
+extern void	rtcl_838x_dram_start(void);
+extern int	rtcl_838x_dram_size;
+
+extern void	(*rtcl_838x_dram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_838x_sram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+
+extern void	(*rtcl_839x_dram_set_rate_mem)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_839x_sram_set_rate_mem)(int clk_idx, int ctrl0, int ctrl1);
+
+extern void	rtcl_839x_dram_start(void);
+extern int	rtcl_839x_dram_size;
+
+extern void	(*rtcl_839x_dram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_839x_sram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
```

Comment:
maybe put these in a header with the same name as your sram.S? makes it clear what and where they are from.

Reply:
Added a comment but will not change that.